### ISSUE TITLE
Render Selecta less intrusively

### ISSUE THREAD

**Issue by rschmitt** [garybernhardt/selecta#90]
Currently, Selecta renders itself according to the following procedure:

1. Upon startup, determine the dimensions of the Terminal window
2. Draw Selecta on the bottom of the screen, using up to 21 lines (20 visible choices plus one prompt)
3. After a selection is made, move the cursor to the bottom of the screen

This works fine for a Command-T replacement, but it's fairly jarring when using Selecta as an extension of the shell. Ideally, summoning Selecta with Ctrl-S should look like this:

1. Selecta renders itself directly beneath the current prompt line, moving the screen down only as far as necessary
2. When a selection is made, Selecta vanishes
3. The cursor moves back to the _original_ prompt line, with the selection added in place, as if the user had typed in the selection manually

I've drawn up a comparison using Heatseeker. The below demos show the following use cases, in order:

1. Writing a selection directly to stdout
2. Saving a selection to an environment variable
3. Appending a selection to the command line using Ctrl-S

Here is Selecta:

![capture](https://cloud.githubusercontent.com/assets/3725049/8273585/49be7604-1829-11e5-8ff7-cb69f633b2be.gif)

Here is Heatseeker:

![capture](https://cloud.githubusercontent.com/assets/3725049/8273576/ecce4226-1828-11e5-9269-e017b54755fa.gif)

The vim integration looks the same in either case, so I did not include it here.

**Comment by garybernhardt** [comment]
Here are the options I see, from easiest to hardest:

1. Draw at the bottom, as Selecta does now. This is easy to implement because the coordinates are absolute.
2. Draw full-height, consuming the whole terminal.
3. Draw from where the cursor currently is, but don't fill the whole screen.
4. Draw from where the cursor currently is *and* save/restore the screen (smcup and rmcup).

It looks like heatseaker is doing (4), unless you're manually clearing afterward. Is that right? How do you calculate coordinates? Or, do you use relative movement ANSI commands?

**Comment by rschmitt** [comment]
I save and restore _the cursor position_ (using [these commands](https://github.com/rschmitt/heatseeker/blob/master/src/ansi.rs#L19-L25)) and combine that with relative cursor movement. And yes, this is trickier on VT100 than it is on the Windows console, because in VT100 the only way of getting the current cursor coordinates is... too complicated to consider, _especially_ since I already know that you don't want to parse terminal escapes in the input.

**Comment by garybernhardt** [comment]
Oh, I see how you did it: you save the cursor position, then restore to it when rendering every frame. That's clever. I was imagining doing it using only relative cursor movements, which seemed very error-prone. With your way, rendering might actually be easier than it is now.

Do you have any opinion about taking up the full screen, vs. the current height of 20 lines? (Or, does anyone else who's watching have an opinion?)

**Comment by rschmitt** [comment]
I really haven't thought that through. It's clear to me that there are some contexts where having more than 20 selections on the screen would be useful. (For instance, I've been finding that heatseeker is sometimes useful as a `less` alternative for man pages and `--help` output, in order to locate interesting command line switches.)

I guess it depends on what you mean by "the full screen." Using all available screen space (from the current cursor position to the bottom, but making space for up to 20 lines if necessary) seems like it would probably be harmless at worst in the terminal, and it wouldn't change the behavior in Vim. It certainly wouldn't be difficult to prototype, although I'm sure I've hardcoded `20` in more than one place.

**Comment by rschmitt** [comment]
Actually, let me amend that. Taking up the full screen is very easy to prototype; it's a one line change in my code. Taking up everything from the cursor to the bottom might not even be _possible_ on VT100 unless you want to use the insane "Query Cursor Position" instruction, or just switch to ncurses. My rendering technique only works if you don't care about the absolute cursor position at any time.

**Comment by louwers** [comment]
This is what it looks like on my machine, and I have to agree: it's not ideal.

![screen shot 2015-12-31 at 3 26 13](https://cloud.githubusercontent.com/assets/649392/12060371/4c05612a-af6e-11e5-9611-3e943805afea.png)

**Comment by garybernhardt** [comment]
I've modified the rendering to fix this, so that Selecta shows up directly below the prompt as you'd expect. It turned out that using relative ANSI motion commands was simpler than what selecta was doing before; I just didn't stumble across the simple solution when thinking about it the first time around. I've pushed the new change to [a branch](/garybernhardt/selecta/tree/less_intrusive_rendering). If it doesn't seem to break anything in the near future, I'll consider it good and merge it into master. The more testing it the merrier, of course!
